src/HttpCacheListener.php

```diff
@@ -273,4 +293,70 @@ public function setVary(Headers $headers)
 
         return $this;
     }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setEtag(HttpRequest $request, HttpResponse $response)
+    {
+        $headers = $response->getHeaders();
+        if (! empty($this->cacheConfig['etag'])
+            && (! $headers->has('etag')
+                || ! empty($this->cacheConfig['etag']['override']))
+        ) {
```

Comment:
I'd invert this condition, and return if you *won't* be setting the ETag. The purpose of the method is for doing exactly that, so it shouldn't be buried in a conditional.